quickjs/src/main/jni/JsMethodProxy.cpp

```diff
@@ -18,53 +18,119 @@
   const jclass methodClass = env->GetObjectClass(method);
 
   const jmethodID getReturnType =
       env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
   const auto returnedClass = static_cast<jclass>(env->CallObjectMethod(method, getReturnType));
 
   const jmethodID isVarArgsMethod = env->GetMethodID(methodClass, "isVarArgs", "()Z");
-  const auto isVarArgs = env->CallBooleanMethod(method, isVarArgsMethod);
+  isVarArgs = false; // TODO: env->CallBooleanMethod(method, isVarArgsMethod);
 
   const jmethodID getParameterTypes =
       env->GetMethodID(methodClass, "getParameterTypes", "()[Ljava/lang/Class;");
   jobjectArray parameterTypes =
       static_cast<jobjectArray>(env->CallObjectMethod(method, getParameterTypes));
   const jsize numArgs = env->GetArrayLength(parameterTypes);
-  // Release any local objects allocated in this frame when we leave this scope.
-  env->PushLocalFrame(numArgs);
-  //argumentLoaders.resize(numArgs);
+  argumentLoaders.resize(numArgs);
   for (jsize i = 0; i < numArgs; ++i) {
     auto parameterType = env->GetObjectArrayElement(parameterTypes, i);
     if (isVarArgs && i == numArgs - 1) {
       const auto parameterClass = env->GetObjectClass(parameterType);
       const jmethodID getComponentType =
           env->GetMethodID(parameterClass, "getComponentType", "()Ljava/lang/Class;");
       parameterType = env->CallObjectMethod(parameterType, getComponentType);
-      //argumentLoaders[i] = typeMap.get(env, static_cast<jclass>(parameterType));
+      argumentLoaders[i] = getJavaToJsConverter(context, static_cast<jclass>(parameterType), false);
       break;
+    } else {
+      argumentLoaders[i] = getJavaToJsConverter(context, static_cast<jclass>(parameterType), true);
     }
-    //  argumentLoaders[i] = typeMap.getBoxed(env, static_cast<jclass>(parameterType));
+    env->DeleteLocalRef(parameterType);
   }
-  env->PopLocalFrame(nullptr);
+  env->DeleteLocalRef(parameterTypes);
 }
 
 jobject JsMethodProxy::call(Context *context, JSValue thisPointer, jobjectArray args) const {
-  JSValue function = JS_GetPropertyStr(context->jsContext, thisPointer, name.c_str());
   const auto totalArgs = std::min<int>(argumentLoaders.size(), context->env->GetArrayLength(args));
   JSValueConst arguments[totalArgs];
   int numArgs;
+  jvalue arg;
   for (numArgs = 0; numArgs < totalArgs && !context->env->ExceptionCheck(); numArgs++) {
-    auto arg = context->env->GetObjectArrayElement(args, numArgs);
+    arg.l = context->env->GetObjectArrayElement(args, numArgs);
     arguments[numArgs] = argumentLoaders[numArgs](context, arg);
+    context->env->DeleteLocalRef(arg.l);
   }
 
   jobject result;
   if (!context->env->ExceptionCheck()) {
-    JSValue callResult = JS_Call(context->jsContext, function, thisPointer, numArgs, arguments);
+    auto property = JS_NewAtom(context->jsContext, name.c_str());
+    JSValue callResult = JS_Invoke(context->jsContext, thisPointer, property, numArgs, arguments);
```

Comment:
I have just noticed that we are not doing that either but the JSValue of the argumentLoaders array might have to be free'd after JS_Invoke(). Also (even it probably doesn't matter much) maybe the ArgumentLoader typedef should return JSValue (instead of JSValueConst).

Reply:
It is... line 140?

Reply:
No, what I mean is that the JSValue instances created by the argument loaders need to be eventually free'd (not relevant for primitives but the reference count needs to be decremented for strings, arrays, objects, ...).

In other words, you probably need to iterate through the "arguments" array after JS_Invoke() and "JS_FreeValue()" them otherwise the JS objects will be leaked.

Also the argument loader lambda should probably return a JSValue (instead of JSValueConst) because it is newly created and it feels weird to free a const value afterwards.

Reply:
```  for (int i = 0; i < numArgs; i++) {
    JS_FreeValue(context->jsContext, arguments[i]);	    JS_FreeValue(context->jsContext, arguments[i]);
  }	  }```

Reply:
I have the arguments as JSValueConst since that's the type the invoke API accepts:  
```
JSValue JS_Invoke(JSContext *ctx, JSValueConst this_val, JSAtom atom,
                  int argc, JSValueConst *argv);
```